test_agent/rust/src/testagent.rs

```diff
@@ -0,0 +1,294 @@
+    listner_map: Vec<String>,
+}
+
+#[async_trait]
+impl UListener for SocketTestAgent {    
+    
+    async fn on_receive(&self, msg: UMessage) {
+        dbg!("OnReceive called");
+
+        let __payload = match &msg.payload.data {
+            Some(data) => {
+                // Now we have access to the Data enum
+                match data {Data::Reference(reference)=>{reference.to_string()}Data::Value(value)=> {let value_str=String::from_utf8_lossy(value);value_str.to_string()},
+    _ => "none".into(), }
+            }
+            None => {
+                println!("No data available");
+                "none".into()
+            }
+        };
+
+        let mut _value:HashMap<String,String> =  HashMap::new();
+        _value.insert("value".into(),__payload.into());
+        let _value_str = serde_json::to_string(&_value).expect("issue in converting to payload");
+       
+        let mut _payload:HashMap<String,String> =  HashMap::new();
+        _payload.insert("payload".into(),_value_str);
+        let _payload_str = serde_json::to_string(&_payload).expect("issue in converting to payload");
+
+        let mut json_message: JsonResponseData = JsonResponseData {
+            action: constants::RESPONSE_ON_RECEIVE.to_owned(),
+            data: HashMap::new(),
+            ue: "rust".to_string(),
+        };
+        
+
+        json_message.data.insert("data".into(),_payload_str);
+       
+        <SocketTestAgent as Clone>::clone(&self)
+            .send_to_tm(json_message)
+            .await;
+    }
+
+    async fn on_error(&self, _err: UStatus) {
+        todo!();
+    }
+}
+
+impl Clone for SocketTestAgent {
+    fn clone(&self) -> Self {
+        SocketTestAgent {
+            utransport: self.utransport.clone(),
+            clientsocket: self.clientsocket.clone(),
+            clientsocket_to_tm:self.clientsocket_to_tm.clone(),
+            listner_map: self.listner_map.clone(),
+        }
+    }
+
+    fn clone_from(&mut self, source: &Self) {
+        *self = source.clone()
+    }
+}
+
+impl SocketTestAgent {
+    pub fn new(test_clientsocket: TcpStreamSync, test_clientsocket_to_tm : TcpStreamSync, utransport: UtransportSocket) -> Self {
+        let socket = Arc::new(Mutex::new(test_clientsocket));
+        let socket_to_tm = Arc::new(Mutex::new(test_clientsocket_to_tm));
+        let clientsocket = socket;
+        let clientsocket_to_tm = socket_to_tm;
+        SocketTestAgent {
+            utransport,
+            clientsocket,
+            clientsocket_to_tm,
+            listner_map: Vec::new(),
+        }
+    }
+
+
+    fn sanitize_input_string(self, input: &str) -> String {
+        input.chars()
+            .map(|c| {
+                match c {
+                    '\x00'..='\x1F' => self.clone().escape_control_character(c),
+                    _ => c.to_string(),
+                }
+            })
+            .collect()
+    }
+     
+    fn escape_control_character(self,c: char) -> String {
```

Comment:
Doesn't use any member fields, so this could be made into a method on the struct, not an instance, i.e.
```suggestion
    fn escape_control_character(c: char) -> String {
```